types/src/jsonrpc/wrapped/batches.rs

```diff
@@ -123,48 +119,50 @@ impl<T> BatchesState<T> {
 		// better performances in real-world situations. More brainstorming and benchmarking could
 		// get helpful here.
 
-		// Because of long-standing Rust lifetime issues
-		// (https://github.com/rust-lang/rust/issues/51526), we can't do this in an elegant way.
-		// If you're reading this code, know that it took several iterations and that I hated my
-		// life while trying to figure out how to make the compiler happy.
-		for batch_id in self.batches.keys().cloned().collect::<Vec<_>>() {
+		for (batch_id, entry) in self.batches.iter_mut().enumerate() {
 			enum WhatCanWeDo {
-				Nothing,
 				ReadyToRespond,
 				Notification(Notification),
 				Request(usize),
 			}
 
 			let what_can_we_do = {
-				let (batch, _) = self.batches.get_mut(&batch_id).expect("all keys are valid; qed");
+				// Unwrap the entry, skipping `None`s
+				let (batch, _) = match entry {
+					Some(entry) => entry,
+					None => continue,
+				};
+
 				let is_ready_to_respond = batch.is_ready_to_respond();
```

Comment:
This call seems to be non-free, and it seems like its result is only used for the `None` case below? If so, maybe this is more efficient:

```
				match batch.next() {
					None => {
						if batch.is_ready_to_respond() {
							WhatCanWeDo::ReadyToRespond
						} else {
							continue;
						}
					}
					Some(batch::BatchInc::Notification(n)) => WhatCanWeDo::Notification(n),
					Some(batch::BatchInc::Request(inner)) => WhatCanWeDo::Request(inner.id()),
				}
```

Reply:
I'm always down to remove boolean flag variables, since they usually are code smell :).